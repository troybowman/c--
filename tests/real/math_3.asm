
.data

  _str0:
    .asciiz " "
    .align 2

  _str1:
    .asciiz "\n"
    .align 2

.text

main:

  # |--------------------------------|
  # |              $a0               |
  # |--------------------------------| sp+24  <-- start of caller's frame
  # |           <padding>            |
  # |--------------------------------| sp+20
  # |              $ra               |
  # |--------------------------------| sp+16
  # |     <minimum 4 arg slots>      |
  # |--------------------------------| sp+0
  la $sp, -24($sp)
  sw $ra, 16($sp)
  sw $a0, 24($sp)

  li $t0, 1
  li $t1, 0
  sllv $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 1
  sllv $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 2
  sllv $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 3
  sllv $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 4
  sllv $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str1
  move $a0, $t0
  jal __print_string
  li $t0, 8
  li $t1, 0
  srlv $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 8
  li $t1, 1
  srlv $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 8
  li $t1, 2
  srlv $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 8
  li $t1, 3
  srlv $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 8
  li $t1, 4
  srlv $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str1
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 0
  and $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 1
  and $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 2
  and $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 3
  and $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 4
  and $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str1
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 0
  or $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 1
  or $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 2
  or $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 3
  or $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 4
  or $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str1
  move $a0, $t0
  jal __print_string
  li $t0, 0
  li $t1, 0
  xor $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  li $t1, 0
  xor $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 2
  li $t1, 1
  xor $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 3
  li $t1, 2
  xor $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 4
  li $t1, 3
  xor $t2, $t0, $t1
  move $a0, $t2
  jal __print_int
  la $t0, _str1
  move $a0, $t0
  jal __print_string
  li $t0, 0
  not $t1, $t0
  move $a0, $t1
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 1
  not $t1, $t0
  move $a0, $t1
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 2
  not $t1, $t0
  move $a0, $t1
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 3
  not $t1, $t0
  move $a0, $t1
  jal __print_int
  la $t0, _str0
  move $a0, $t0
  jal __print_string
  li $t0, 4
  not $t1, $t0
  move $a0, $t1
  jal __print_int
  la $t0, _str1
  move $a0, $t0
  jal __print_string

__leavemain:
  lw $a0, 24($sp)
  lw $ra, 16($sp)
  la $sp, 24($sp)
  jal __exit

__print_string:
  li $v0, 4
  syscall
  jr $ra

__print_int:
  li $v0, 1
  syscall
  jr $ra

__print_char:
  li $v0, 11
  syscall
  jr $ra

__print_hex:
  li $v0, 34
  syscall
  jr $ra

__exit:
  li $v0, 10
  syscall
  jr $ra
