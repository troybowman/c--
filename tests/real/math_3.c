extern void printf(char format[], ...);

void main(void)
{
  printf("%d ",  1 << 0);
  printf("%d ",  1 << 1);
  printf("%d ",  1 << 2);
  printf("%d ",  1 << 3);
  printf("%d\n", 1 << 4);

  printf("%d ",  8 >> 0);
  printf("%d ",  8 >> 1);
  printf("%d ",  8 >> 2);
  printf("%d ",  8 >> 3);
  printf("%d\n", 8 >> 4);

  printf("%d ",  (1 & 0));
  printf("%d ",  (1 & 1));
  printf("%d ",  (1 & 2));
  printf("%d ",  (1 & 3));
  printf("%d\n", (1 & 4));

  printf("%d ",  (1 | 0));
  printf("%d ",  (1 | 1));
  printf("%d ",  (1 | 2));
  printf("%d ",  (1 | 3));
  printf("%d\n", (1 | 4));

  printf("%d ",  (0 ^ 0));
  printf("%d ",  (1 ^ 0));
  printf("%d ",  (2 ^ 1));
  printf("%d ",  (3 ^ 2));
  printf("%d\n", (4 ^ 3));

  printf("%d ",  ~0);
  printf("%d ",  ~1);
  printf("%d ",  ~2);
  printf("%d ",  ~3);
  printf("%d\n", ~4);
}
