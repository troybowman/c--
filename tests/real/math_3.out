1 2 4 8 16
8 4 2 1 0
0 1 0 1 0
1 1 3 3 5
0 1 3 1 7
-1 -2 -3 -4 -5
